reference-implementation/lib/readable-stream.js

```diff
@@ -86,273 +89,126 @@ class ReadableStream {
+          return ReadableStreamDefaultReaderRead(reader).then(({ value, done }) => {
+            if (done === true) {
+              return undefined;
+            }
+
+            currentWrite = WritableStreamDefaultWriterWrite(writer, value);
+            return currentWrite;
+          });
+        })
+        .then(pipeLoop);
       }
 
-      abortWriterCancelReader(reason, preventAbort, preventCancel);
-    }
-
-    function handleReadValue(value) {
-      // console.log('pipeTo(): handleReadValue()');
-
-      _lastWrite = _writer.write(value);
-      _lastWrite.catch(handleWriteRejection);
-
-      // dest may be already errored. But proceed to write().
-      _allWrites = Promise.all([_allWrites, _lastWrite]);
-
-      doPipe();
-    }
-
-    function handleReadDone() {
-      // console.log('pipeTo(): handleReadDone()');
-
-      // Does not need to wait for lastRead since it occurs only on source closed.
-
-      releaseReader();
-
-      if (preventClose === false) {
-        // console.log('pipeTo(): Close dest');
+      pipeLoop().catch(err => {
+        currentWrite = Promise.resolve();
+        rethrowAssertionErrorRejection(err);
+      });
+
+      // Errors must be propagated forward
+      reader._closedPromise.catch(storedError => {
+        if (preventAbort === false) {
+          performAsyncShutdown(() => WritableStreamAbort(dest, storedError), true, storedError);
+        } else {
+          performShutdown(true, storedError);
+        }
+      });
+
+      // Errors must be propagated backward
+      writer._closedPromise.catch(storedError => {
+        if (preventCancel === false) {
+          performAsyncShutdown(() => ReadableStreamCancel(this, storedError), true, storedError);
+        } else {
+          performShutdown(true, storedError);
+        }
+      });
+
+      // Closing must be propagated forward
+      reader._closedPromise.then(() => {
+        if (preventClose === false) {
+          performAsyncShutdown(() => WritableStreamDefaultWriterCloseWithErrorPropagation(writer));
+        } else {
+          performShutdown();
+        }
+      });
 
-        // We don't use writer.closed. We can ensure that the microtask for writer.closed is run before any
-        // writer.close() call so that we can determine whether the closure was caused by the close() or ws was already
-        // closed before pipeTo(). It's possible but fragile.
-        _writer.close().then(
-          () => {
-            return _allWrites;
-          },
-          reason => {
-            releaseWriter();
-            finishWithRejection(reason);
-          }
-        ).then(
-          () => {
-            releaseWriter();
-            finishWithFulfillment();
-          }
-        );
-        _state = 'closingDest';
+      // Closing must be propagated backward
+      writer._closedPromise.then(() => {
+        const destClosed = new TypeError('the destination writable stream closed before all data could be piped to it');
```

Comment:
Since this can only happen if running pipeTo on a destination that was already closed beforehand, maybe this error should be more specific and say that's what happened? This can't possibly happen in the middle of piping, can it?